federation/pkg/controller/service/servicecontroller.go

```diff
@@ -0,0 +1,956 @@
+			//if err != nil {
+			//	return fmt.Errorf("Error getting DNS records for service %s: %v", namespacedName, err), retryable
+			//}
+			//if !exists {
+			//	needDelete = false
+			//}
+		}
+
+		if needDelete {
+			glog.Infof("Deleting existing DNS records for service %s that no longer needs them.", key)
+			s.eventRecorder.Event(service, api.EventTypeNormal, "DeletingDNSRecords", "Deleting DNS records")
+			//todo:Jesse
+			//if err := s.dns.EnsureDNSRecordsDeleted(service); err != nil {
+			//	return err, retryable
+			//}
+			s.eventRecorder.Event(service, api.EventTypeNormal, "DeletedDNSRecords", "Deleted DNS records")
+		}
+
+		service.Status.LoadBalancer = v1.LoadBalancerStatus{} // TODO: quinton: is the DNS name in here?
+	} else {
+		glog.V(2).Infof("Ensuring DNS records for service %s", key)
+
+		// TODO: We could do a dry-run here if wanted to avoid the spurious cloud-calls & events when we restart
+
+		// The DNS records don't exist yet, so create them.
+		// TODO: quinton: Split this up and handle each record separately.
+		//                Beware though, creating distant ones before near ones can lead to race condidtions on name resolution.
+		//                (e.g. client local resolution fails as DNS record doesn't exist yet, but then tries remote resolution and that succeeds,
+		//                so it gets stuck for a while on a remote service shard.
+		//                Solutions: either use transactions (where providers suport them, e.g. Google Cloud DNS) or just make
+		//                sure that the creation sequence is zone-local, region-local, sub-continent local etc.  The latter probably
+		//                works for all cases, so transactions are probably not necessary.
+		//                Actually, on further thought, the race condition exists anyway - neither of the above tow really solve it.
+		//                Ho hum, have to think harder.
+		s.eventRecorder.Event(service, api.EventTypeNormal, "CreatingDNSRecords", "Creating DNS Records")
+		err := s.ensureDNSRecords(clusterName, service)
+		if err != nil {
+			return fmt.Errorf("Failed to create DNS records for service %s: %v", key, err), retryable
+		}
+		s.eventRecorder.Event(service, api.EventTypeNormal, "CreatedDNSRecords", "Created DNS records")
+	}
+
+	serviceInMap, ok := s.cache.get(service.Name)
+	if !ok {
+		return fmt.Errorf("service cache lost")
+	}
+	// TODO: remove dup ip info?
+	serviceInMap.appliedState.Spec.ExternalIPs = append(serviceInMap.appliedState.Spec.ExternalIPs, service.Spec.ExternalIPs)
+	// write back to federation apiserver
+
+	// TODO to check LB info equals
+	//if !api.LoadBalancerStatusEqual(previousState, &service.Status.LoadBalancer) { // TODO: quinton: only interested in the DNS record here.
+	//	// Not even sure it needs updating - check.
+	//} else {
+	_, err := s.federationClient.Core().Services().Update(serviceInMap)
+	if err != nil {
+		return fmt.Errorf("fail to update federation service")
```

Comment:
Need to add service name, and error in output.   Or better still, return the original error (err) so that people debugging can see what went wrong.